Sol.java

```diff
@@ -0,0 +1,104 @@
+import java.util.*;
+
+public class Sol {
+    int[] in ;
+    HashMap < Integer, Integer > exp;
+    LinkedList < Move > moves_coll;
+    int zeropos;
+
+    public LinkedList<Move> move_cars(int[] in , int[] out) {
+        initialise( in , out);
+        solve();
+      
+        return moves_coll;
+    }
+
+    private void initialise(int[] in , int[] out) {
+        exp = new HashMap < Integer, Integer > ();
+        moves_coll = new LinkedList < Move > ();
+        this.in = in ;
```

Comment:
The think is that it's not a copy. In the method  `move` you have lines:  in [di] = val; in [i] = 0; it means you change the content of that array. And it's the same array as you pass to `move_cars`. This `in` array was created in some outer function (let's say it's name is `outerFunction`) that calls `move_cars`. After calling it  `in` array in `outerFunction` is changed. which is not good. Do you repair somewhere that array? I didn't notice it.